rs/src/connections/io.rs

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+use std::task::Poll;
+
+/// Helper used when converting Future interfaces to poll-based interfaces.
+/// Stores excess data that can be reused on future polls.
+#[derive(Default)]
+pub(crate) struct ReadBuffer(Option<Vec<u8>>);
+
+impl ReadBuffer {
+    /// Removes any data stored in the read buffer
+    pub fn take_data(&mut self) -> Option<Vec<u8>> {
+        self.0.take()
+    }
+
+    /// Writes as many bytes as possible to the readbuf, stashing any extra.
+    pub fn put_data(
+        &mut self,
+        target: &mut tokio::io::ReadBuf<'_>,
+        bytes: &[u8],
```

Comment:
It might make sense to take a `Message` by value here and move the `Message` from the caller. Doing that, you'd then put `Message` in `ReadBuffer` instead of the `Vec<u8>` and hat way, you don't have to copy the part of the message that didn't send on line 34. If the message is very large and the `ReadBuf` is very small, you could end up copying the same tail of the message over and over in the `AsyncRWWebSocket::poll_read` call site. 

This suggestion is really just about the perf aspect; I think the design here is fine so please ignore until you're working on perf.